rfcs/0000-question-sets.md

```diff
@@ -0,0 +1,150 @@
+
+# Potential future enhancements
+
+The following features are not necessary for an MVP version of this feature, but are called out for the sake of discussion.
+
+## Question deprecation
+
+To ensure safety guarantees, once a question in a question set exists, it cannot be renamed or deleted. However, as question sets eveolve over the course of many years, certain questions may become outdated or broken and there may not be sufficient resources to update them. To make it easy for this to be communicated to consumers of questions, a "deprecated" flag could be added to a question's `info.json` file:
+
+```json
+{
+  "uuid": "754dac56-fc1f-4888-a74a-1b365fd48376",
+  "deprecated": true
+}
+```
+
+Deprecated questions could render a warning in instructor views of the question, on a question's row in an assessment overview, or in other useful locations across PrairieLearn.
+
+# Unresolved questions and concerns
+
+## Feature name
+
+With the existence of assessment sets, "set" might become too much of an overloaded term if used for this feature. A name like "question collections" may be more appropriate.
+
+## Question set IDs
+
+Question set IDs are how question sets are referenced in assessments. Ideally, question set IDs should have the following properties:
+
+* Static (do not change over time)
+* Consistent across instances of PrairieLearn
+* Human-readable
+
+Because they must be human readable, a question set's UUID is unsuitable as the user-facing question set ID. The `name` value in `infoCourse.json` could be repurposed for this, but it should have additional restrictions (no spaces) and has a different semantic meaning. A new `id` value could be introduced that has no meaning in standard courses.
+
+## Access control within a PrairieLearn instance
+
+As with normal courses, access to question sets should be carefully controlled. It should also be consistent: if I'm staff for a course instance, I should be able to access question sets that are referenced by my courses' assignments. It's not clear how this access should be requested, granted, or managed.
+
+These permissions should likely not live in a question set's repo - they're liekly going to be specific to the instance of PrairieLearn a course is running on.
+
+Perhaps permissions should exist at the course level - a PrairieLearn admin can grant course X the ability to use and view question set Y. That permission would be transitive: if I'm course staff for course X, and course X has the ability to use question set Y, then I should be able to view and use anything in question set Y.
```

Comment:
I think the risk with transitive permissions is that some child courses might have very large and permissive course staffs. I don't really know what to do about this 😄

Reply:
I wasn't sure about this initially either. There are definitely some security risks with this, but I don't see how to prevent an insane amount of complexity or unexpected behavior without this transitive model. There are a lot of valid use cases for your standard member of course staff accessing all questions (including shared ones):

* Building an exam
* Reviewing an exam prior to launching it
* Reviewing student-submitted issues on an assessment
* Helping a student in office hours

The only case where I really see problems here is if a) a student is course staff for X, b) a student is taking course Y, and c) X and Y both use the same set of shared questions. That seems unlikely to me, so I'm willing to overlook it, tbh 🙂

Reply:
I agree this is a horrible can of worms to try and lock it down further.

Do shared questions appear in the big list of questions within a child course?

Reply:
I don't think they should, but I do think a separate view like that would be helpful.

Reply:
I can see this both ways. Sometimes I'd want to see only the questions that are actually local to the current course, but when I'm are trying to find another question to include on the homework, I'd like to see everything available in a big list that I can search.